autonav_ws/src/autonav_hardware/src/can_node.py

```diff
@@ -0,0 +1,289 @@
+import struct
+from ctypes import Structure, c_bool, c_uint8
+
+arbitration_ids = {
+    "EStop": 0,
+    "MobilityStop": 1,
+    "MobilityStart": 9,
+    "MotorsCommand": 10,
+    "SafetyLightsCommand": 13,
+    "OdometryFeedback": 14,
+    "ObjectDetection": 20,
+    "LinearPIDStatistics": 50,
+    "AngularPIDStatistics": 51,
+    "MotorStatisticsFrontMotors": 52,
+    "MotorStatisticsBackMotors": 53,
+    "ConbusLowerBound": 1000,
+    "ConbusUpperBound": 1400
+}
+
+class SafetyLightsPacket(Structure):
+    _fields_ = [
+        ("autonomous", c_bool, 1),
+        ("mode", c_uint8, 7),
+        ("brightness", c_uint8, 8),
+        ("red", c_uint8, 8),
+        ("green", c_uint8, 8),
+        ("blue", c_uint8, 8),
+        ("blink_period", c_uint8, 8)
+    ]
+
+class CanConfig:
+    def __init__(self):
+        self.odom_feedback_scaler = 10000
+        self.linear_pid_scaling_factor = 1000
+        self.angular_pid_scaling_factor = 1000
+
+
+class can_node(Node):
+    def __init__(self):
+        super.__init__("CAN_node")
+        self.write_config(CanConfig)
+
+        # can
+        self.can = None
+
+        # safety lights
+        self.safetyLightsSubscriber = self.create_subscription(
+            SafetyLights,
+            "autonav/safety_lights",
+            self.on_safety_lights_received(),
+            20
+        )
+
+        # motor messages
+        self.motorInputSubscriber = self.create_subscription(
+            MotorInput,
+            "autonav/motor_input",
+            self.on_motor_input_received(),
+            20
+        )
+
+        self.motorFeedbackPublisher = self.create_publisher(
+            MotorFeedback,
+            "/autonav/motor_feedback", 
+            20
+        )
+
+        # conbus
+        self.conbusSubscriber = self.create_subscription(
+            Conbus,
+            "/autonav/conbus/instruction", 
+            self.on_conbus_received(), 
+            20
+        )
+
+        self.conbusPublisher = self.create_publisher(
+            Conbus, 
+            "/autonav/conbus/data", 
+            20
+        )
+
+        # motor statistics and PID tuning
+        self.linearPIDStatisticsPublisher = self.create_publisher(
+            LinearPIDStatistics,
+            "autonav/linear_pid_statistics",
+            20
+        )
+        self.angularPIDStatisticsPublisher = self.create_publisher(
+            AngularPIDStatistics,
+            "autonav/angular_pid_statistics",
+            20
+        )
+        self.motorStatisticsFrontMotorsPublisher = self.create_publisher(
+            MotorStatisticsFrontMotors,
+            "autonav/motor_statistics_front_motors",
+            20
+        )
+        self.motorStatisticsBackMotorsPublisher = self.create_publisher(
+            MotorStatisticsBackMotors,
+            "autonav/motor_statistics_back_motors",
+            20
+        )
+
+
+    def init(self):
+        # can threading
+        self.canTimer = self.create_timer(0.5, self.can_worker)
```

Comment:
is the .5 a timeout and should it be shorter or longer? half a second seems slow for CAN stuff but idk what the number means

Reply:
^ this is attempting to make a new can worker if there isn't one already every .5 seconds, seems reasonable